demos/speculation_over_syscall.cc

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __linux__
+#  error Unsupported OS. Linux required.
+#endif
+
+#ifndef __aarch64__
+#  error Unsupported architecture. ARM64 required.
+#endif
+
+#include <array>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+
+#include <signal.h>
+#include <sys/syscall.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+
+const char *public_data = "Hello, world!";
+const char *private_data = "It's a s3kr3t!!!";
+
+// Local handler necessary for avoiding local/global linking mismatches on ARM.
+// When we use extern char[] declaration for a label defined in assembly, the
+// compiler yields this sequence that fails loading the actual address of the
+// label:
+// adrp x0, :got:label
+// ldr x0, [x0, #:got_lo12:label]
+// On the other hand when we use this local handler, the compiler yield this
+// sequence of instructions:
+// adrp x0, label
+// add x0, x0, :lo12:label
+// and that works correctly because it if an effective equivalent of
+// adr x0, label.
+static void local_handler() {
+  asm volatile("b afterspeculation");
+}
+
+static char leak_byte(const char *data, size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &isolated_oracle = sidechannel.GetOracle();
+
+  for (int run = 0;; ++run) {
+    size_t safe_offset = run % strlen(public_data);
+    sidechannel.FlushOracle();
+
+    // Architecturally access the safe offset.
+    ForceRead(isolated_oracle.data() + static_cast<size_t>(data[safe_offset]));
+
+    // Sends a SIGUSR1 signal to itself. The signal handler shifts the control
+    // flow to the "afterspeculation" label.
```

Comment:
We have a good reason not to use the `syscall` library function here -- because it brings Spectre V1 into scope -- but readers might reasonably wonder and we should leave a comment about that.

Reply:
Done.